### PR TITLE
0.2-dev: LF-RFID emulation support

### PR DESCRIPTION
**Pull request by twitchyliquid64**
Could use some advice on appropriate types, in particular what to do about `PROTOCOLS_MAX`.

I was able to open the front door of my apartment complex with this code:

```rust
let dict = protocol_dict_alloc(&LFRFID_PROTOCOLS, PROTOCOLS_MAX);
let protocol_id = protocol_dict_get_protocol_by_name(dict, sys::c_string!("H10301"));

let mut data: [u8; 3] = [<REDACTED>];
protocol_dict_set_data(dict, protocol_id, data.as_ptr(), 3);

let rf_worker = lfrfid_worker_alloc(dict);
lfrfid_worker_start_thread(rf_worker);
lfrfid_worker_emulate_start(rf_worker, protocol_id);

// wait for user input

lfrfid_worker_stop(rf_worker);
lfrfid_worker_stop_thread(rf_worker);
lfrfid_worker_free(rf_worker);
protocol_dict_free(dict);
```

### CODE REVIEW

**Comment by dcoles** [comment]
> I was able to open the front door of my apartment complex with this code:

I've been having a lot of fun seeing what I can emulate. So far I've managed to use it for my RFID badge and my car fob (sadly it only works once, probably to prevent replay attacks).

**Comment by twitchyliquid64** [comment]
Yeah its pretty dope! Its been a long while since ive seen a hackable device with a reasonable 125kHz frontend (basically the first since the proxmark3).

**Comment by dcoles** [comment]
> I was able to open the front door of my apartment complex with this code:

We should probably move some of these into an examples directory, since they're super helpful to have.

**Comment by twitchyliquid64** [comment]
I plan to work on an actually-useful app that'l drive the APIs im implementing :)

File: crates/sys/src/lfrfid.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+use crate::opaque;
+
+
+
+// The global array of all protocols implemented by the firmware.
+opaque!(ProtocolsArray);
+
+// The number of entries in LFRFID_PROTOCOLS
+pub const PROTOCOLS_MAX: usize = 10; // TODO: find some way to keep sync with LFRFIDProtocolMax
```
Comment (dcoles) [inline crates/sys/src/lfrfid.rs:13]: It should be fine if this doesn't match `flipperzero-firmware`.

If `PROTOCOLS_MAX < LFRFIDProtocolMax`, this just means we can't access the new protocols.
If `PROTOCOLS_MAX > LFRFIDProtocolMax` (old firmware), programs will still work, but will [trigger a `furi_crash` if the application tries to use a too-new protocol](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware@0f9ea92/-/blob/lib/toolbox/protocols/protocol_dict.c?L76).

If it becomes particularly problematic, we can probably request a `extern` variable with the correct `LFRFIDProtocolMax` value.

**Review by dcoles — CHANGES_REQUESTED** [review]

File: crates/sys/src/lfrfid.rs (hunk 1 above)
Comment (twitchyliquid64) [reply crates/sys/src/lfrfid.rs:13]: Sounds good. Hey @hedger any chance we could get `LFRFIDProtocolMax` as an API symbol? ^^'

File: crates/sys/src/lfrfid.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    pub fn lfrfid_worker_free(worker: *mut Worker);
+    #[link_name="lfrfid_worker_start_thread"]
+    pub fn lfrfid_worker_start_thread(worker: *mut Worker);
+    #[link_name="lfrfid_worker_stop_thread"]
+    pub fn lfrfid_worker_stop_thread(worker: *mut Worker);
+
+    #[link_name="lfrfid_worker_emulate_start"]
+    pub fn lfrfid_worker_emulate_start(worker: *mut Worker, protocol_id: ProtocolID);
+    #[link_name="lfrfid_worker_stop"]
+    pub fn lfrfid_worker_stop(worker: *mut Worker);
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply crates/sys/src/lfrfid.rs]: Done.

File: crates/sys/src/lfrfid.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+opaque!(ProtocolsArray);
+
+// The number of entries in LFRFID_PROTOCOLS
+pub const PROTOCOLS_MAX: usize = 10; // TODO: find some way to keep sync with LFRFIDProtocolMax
+
+// The runtime instantiation of a protocol dictionary.
+opaque!(Dict);
+
+// Corresponds to ProtocolId in the firmware source.
+pub type ProtocolID = usize;
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply crates/sys/src/lfrfid.rs]: Done.

File: crates/sys/src/lfrfid.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    pub static LFRFID_PROTOCOLS: ProtocolsArray;
+
+    #[link_name="protocol_dict_alloc"]
+    pub fn protocol_dict_alloc(protocols: &'static ProtocolsArray, num_protocols: usize) -> *mut Dict;
+    #[link_name="protocol_dict_free"]
+    pub fn protocol_dict_free(dict: *mut Dict);
+    #[link_name="protocol_dict_get_protocol_by_name"]
+    pub fn protocol_dict_get_protocol_by_name(dict: *mut Dict, name: *const c_char) -> ProtocolID;
+    #[link_name="protocol_dict_set_data"]
+    pub fn protocol_dict_set_data(dict: *mut Dict, protocol_id: ProtocolID, data: *const u8, data_len: usize);
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply crates/sys/src/lfrfid.rs]: Done.

**Comment by dcoles** [comment]
@twitchyliquid64 Ah! Sorry. I merged and deleted the `0.2-dev` branch and it closed this PR.

This branch is now identical to `main`.

**Review by dcoles — APPROVED** [review]

**Comment by twitchyliquid64** [comment]
My git fu is terrible but i think it made a new `0.2-dev` branch and merged it there, rather than into `main` ?

**Comment by twitchyliquid64** [comment]
Nvm its now in main i think

**Comment by dcoles** [comment]
> My git fu is terrible but i think it made a new `0.2-dev` branch and merged it there, rather than into `main` ?

Just me doing some `git` shenanigans. I had to first merge the PR into `0.2-dev` and then fast-forward merge it to `main`.

**Pull request merged**